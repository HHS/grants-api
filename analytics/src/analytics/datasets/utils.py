"""Stores utility functions for Dataset classes."""

import json


def load_json_file(path: str) -> list[dict]:
    """Load contents of a JSON file into a dictionary."""
    with open(path) as f:
        return json.load(f)


def dump_to_json(path: str, data: dict | list[dict]) -> None:
    """Write a dictionary or list of dicts to a json file."""
    with open(path, "w") as f:
        # Uses ensure_ascii=False to preserve emoji characters in output
        # https://stackoverflow.com/a/52206290/7338319
        json.dump(data, f, indent=2, ensure_ascii=False)
