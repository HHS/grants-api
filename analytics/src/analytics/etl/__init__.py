"""Run extract, transform, and load (ETL) pipelines to generate datasets."""
