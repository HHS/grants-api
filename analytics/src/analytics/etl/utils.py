"""Expose utility functions used across the etl sub-package."""

import json
from pathlib import Path
from typing import TypeVar

from pydantic import BaseModel

# Define a generic type variable for the schema
T = TypeVar("T", bound=BaseModel)


def load_config(config_path: Path, schema: type[T]) -> T:
    """
    Load a JSON config file and validate it against a Pydantic schema.

    Parameters
    ----------
    config_path:
        Path to the JSON config file.
    schema:
        The Pydantic schema class to validate the JSON data.

    Returns
    -------
    An instance of the schema containing validated config data.

    """
    with open(config_path) as f:
        config_data = json.load(f)

    return schema(**config_data)
