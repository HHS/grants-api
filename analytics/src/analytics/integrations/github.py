"""Integrate with GitHub to read and write data from projects and repos."""

import shlex
import subprocess
from pathlib import Path

# Set the max number of records to return with CLI commands to 10,000
# NOTE: GitHub exports data in batches of 100 so exporting 10k issues could take over a minute
# TODO(@widal001): 2023-11-29 - Switch to incremental export pattern
#   related issue: https://github.com/HHS/simpler-grants-gov/issues/775
MAX_RECORDS = 10000


def pipe_command_output_to_file(command: str, output_file: str) -> None:
    """Write the command line output to a file."""
    # make sure the output file's directory exists
    file_path = Path(output_file)
    file_path.parent.mkdir(exist_ok=True, parents=True)
    # invoke the command via a subprocess and write the output to a file
    with open(output_file, "w", encoding="utf-8") as f:
        subprocess.call(shlex.split(command), stdout=f)  # noqa: S603


def export_project_data(owner: str, project: int, output_file: str) -> None:
    """Export and write GitHub project data to a JSON file."""
    print(f"Exporting project data from {owner}/{project} to {output_file}")
    command = (
        f"gh project item-list {project} --format json --owner {owner} -L {MAX_RECORDS}"
    )
    pipe_command_output_to_file(command, output_file)


def export_issue_data(owner: str, repo: str, output_file: str) -> None:
    """Export and write GitHub issue data to a JSON file."""
    print(f"Exporting issue data from {owner}/{repo} to {output_file}")
    command = (
        f"gh issue list --json number,createdAt,closedAt,labels,title "
        f"-R {owner}/{repo} -L {MAX_RECORDS} --state all"
    )
    pipe_command_output_to_file(command, output_file)
