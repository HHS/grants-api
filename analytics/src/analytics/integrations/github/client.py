"""Expose a client for making calls to GitHub's GraphQL API."""

import logging
from typing import Any

import requests

from config import get_db_settings

logger = logging.getLogger(__name__)


class GraphqlError(Exception):
    """
    Exception raised for errors returned by the GraphQL API.

    Attributes
    ----------
    errors : list
        List of error details returned by the API.
    message : str
        Human-readable explanation of the error.

    """

    def __init__(self, errors: list[dict]) -> None:
        """Initialize the GraphqlError."""
        self.errors = errors
        self.message = f"GraphQL API returned errors: {errors}"
        super().__init__(self.message)


class GitHubGraphqlClient:
    """
    A client to interact with GitHub's GraphQL API.

    Methods
    -------
    execute_paginated_query(query, variables, data_path, batch_size=100)
        Executes a paginated GraphQL query and returns all results.

    """

    def __init__(self) -> None:
        """
        Initialize the GitHubClient.

        Parameters
        ----------
        token : str
            GitHub personal access token for authentication.

        """
        settings = get_db_settings()
        self.endpoint = "https://api.github.com/graphql"
        self.headers = {
            "Authorization": f"Bearer {settings.github_token}",
            "Content-Type": "application/json",
            "GraphQL-Features": "sub_issues,issue_types",
        }

    def execute_query(self, query: str, variables: dict[str, str | int]) -> dict:
        """
        Make a POST request to the GitHub GraphQL API.

        Parameters
        ----------
        query : str
            The GraphQL query string.
        variables : dict
            A dictionary of variables to pass to the query.

        Returns
        -------
        dict
            The JSON response from the API.

        """
        response = requests.post(
            self.endpoint,
            headers=self.headers,
            json={"query": query, "variables": variables},
            timeout=60,
        )
        response.raise_for_status()
        result = response.json()
        if "errors" in result:
            raise GraphqlError(result["errors"])
        return result

    def execute_paginated_query(
        self,
        query: str,
        variables: dict[str, Any],
        path_to_nodes: list[str],
        batch_size: int = 100,
    ) -> list[dict]:
        """
        Execute a paginated GraphQL query.

        Parameters
        ----------
        query : str
            The GraphQL query string.
        variables : dict
            A dictionary of variables to pass to the query.
        path_to_nodes : list of str
            The path to traverse the response data to extract the "nodes" list,
            so the nodes can be combined from multiple paginated responses.
        batch_size : int, optional
            The number of items to fetch per batch, by default 100.

        Returns
        -------
        list of dict
            The combined results from all paginated responses.

        """
        all_data = []
        has_next_page = True
        variables["batch"] = batch_size
        variables["endCursor"] = None

        while has_next_page:
            response = self.execute_query(query, variables)
            data = response["data"]

            # Traverse the data path to extract nodes
            for key in path_to_nodes:
                data = data[key]

            all_data.extend(data["nodes"])

            # Handle pagination
            page_info = data["pageInfo"]
            has_next_page = page_info["hasNextPage"]
            variables["endCursor"] = page_info["endCursor"]

        return all_data
