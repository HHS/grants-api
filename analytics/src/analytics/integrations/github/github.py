"""Expose a client for making calls to GitHub's GraphQL API."""

import functools
import logging
from typing import Any, Callable

import requests

from config import get_db_settings

logger = logging.getLogger(__name__)


class GraphqlError(Exception):
    """
    Exception raised for errors returned by the GraphQL API.

    Attributes
    ----------
    errors : list
        List of error details returned by the API.
    message : str
        Human-readable explanation of the error.

    """

    def __init__(self, errors: list[dict]) -> None:
        """Initialize the GraphqlError."""
        self.errors = errors
        self.message = f"GraphQL API returned errors: {errors}"
        super().__init__(self.message)


def github_api_error_handler(api_call: Callable) -> Callable:
    """
    Wrap GitHub API calls with error handling.

    Parameters
    ----------
    api_call : Callable
        The GitHub API call function to wrap.

    Returns
    -------
    Callable
        A wrapped function that catches and logs errors.

    """

    @functools.wraps(api_call)
    def try_to_make_api_call_and_catch_error(
        *args: Any,  # noqa: ANN401
        **kwargs: Any,  # noqa: ANN401
    ) -> dict | None:
        try:
            return api_call(*args, **kwargs)
        except Exception:
            logger.exception("GitHub API error")
            return None

    return try_to_make_api_call_and_catch_error


class GitHubGraphqlClient:
    """
    A client to interact with GitHub's GraphQL API.

    Methods
    -------
    execute_paginated_query(query, variables, data_path, batch_size=100)
        Executes a paginated GraphQL query and returns all results.

    """

    def __init__(self) -> None:
        """
        Initialize the GitHubClient.

        Parameters
        ----------
        token : str
            GitHub personal access token for authentication.

        """
        settings = get_db_settings()
        self.endpoint = "https://api.github.com/graphql"
        self.headers = {
            "Authorization": f"Bearer {settings.github_token}",
            "Content-Type": "application/json",
            "GraphQL-Features": "sub_issues,issue_types",
        }

    def execute_query(self, query: str, variables: dict[str, str | int]) -> dict:
        """
        Make a POST request to the GitHub GraphQL API.

        Parameters
        ----------
        query : str
            The GraphQL query string.
        variables : dict
            A dictionary of variables to pass to the query.

        Returns
        -------
        dict
            The JSON response from the API.

        """
        response = requests.post(
            self.endpoint,
            headers=self.headers,
            json={"query": query, "variables": variables},
            timeout=60,
        )
        response.raise_for_status()
        result = response.json()
        if "errors" in result:
            raise GraphqlError(result["errors"])
        return result

    def execute_paginated_query(
        self,
        query: str,
        variables: dict[str, Any],
        path_to_nodes: list[str],
        batch_size: int = 100,
    ) -> list[dict]:
        """
        Execute a paginated GraphQL query.

        Parameters
        ----------
        query : str
            The GraphQL query string.
        variables : dict
            A dictionary of variables to pass to the query.
        path_to_nodes : list of str
            The path to traverse the response data to extract the "nodes" list,
            so the nodes can be combined from multiple paginated responses.
        batch_size : int, optional
            The number of items to fetch per batch, by default 100.

        Returns
        -------
        list of dict
            The combined results from all paginated responses.

        """
        all_data = []
        has_next_page = True
        variables["batch"] = batch_size
        variables["endCursor"] = None

        while has_next_page:
            response = self.execute_query(query, variables)
            data = response["data"]

            # Traverse the data path to extract nodes
            for key in path_to_nodes:
                data = data[key]

            all_data.extend(data["nodes"])

            # Handle pagination
            page_info = data["pageInfo"]
            has_next_page = page_info["hasNextPage"]
            variables["endCursor"] = page_info["endCursor"]

        return all_data
