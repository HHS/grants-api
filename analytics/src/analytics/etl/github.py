"""Extract and transform GitHub data to create the `GitHubIssues` dataset."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from pathlib import Path

import pandas as pd
from pydantic import BaseModel, ValidationError

from analytics.datasets.issues import (
    GitHubIssues,
    IssueMetadata,
    IssueType,
)
from analytics.datasets.utils import load_json_file
from analytics.integrations import github

logger = logging.getLogger(__name__)

# ========================================================
# Config interfaces
# ========================================================


@dataclass
class InputFiles:
    """Expected input files for loading this dataset."""

    roadmap: str
    sprint: str


class GitHubProjectConfig(BaseModel):
    """Configurations for GitHub projects ETL."""

    roadmap_project: RoadmapConfig
    sprint_projects: list[SprintBoardConfig]
    temp_dir: str = "data"
    output_file: str = "data/delivery-data.json"


class RoadmapConfig(BaseModel):
    """
    Configuration details for a roadmap project.

    Attributes
    ----------
    owner
        The GitHub login for the project owner.
    project_number
        The number of the project.
    quad_field
        The name of the project field that stores the Quad value for a deliverable.
    pillar_field
        The name of the project field that stores the Pillar value for a deliverable.

    """

    owner: str
    project_number: int
    quad_field: str = "Quad"
    pillar_field: str = "Pillar"


class SprintBoardConfig(BaseModel):
    """
    Configuration details for a sprint board project.

    Attributes
    ----------
    owner
        The GitHub login for the project owner.
    number
        The number of the project.
    points_field
        The name of the project field that stores story points or estimates.
    sprint_field
        The name of the project field that stores the sprint value.

    """

    owner: str
    project_number: int
    points_field: str = "Points"
    sprint_field: str = "Sprint"


# ========================================================
# ETL pipeline
# ========================================================


class GitHubProjectETL:
    """Manage the ETL pipeline for GitHub project data."""

    def __init__(self, config: GitHubProjectConfig) -> None:
        """Initialize and run the ETL pipeline for Github project data."""
        # Store the config
        self.config = config
        # Declare private attributes shared across ETL steps
        self._transient_files: list[InputFiles]
        self.client = github.GitHubGraphqlClient()
        self.dataset: GitHubIssues

    def run(self) -> None:
        """Run the ETL pipeline."""
        self.extract()
        self.transform()
        self.load()

    def extract(self) -> None:
        """Run the extract step of the ETL pipeline."""
        temp_dir = Path(self.config.temp_dir)

        # Export the roadmap data
        roadmap_file = str(temp_dir / "roadmap-data.json")
        roadmap = self.config.roadmap_project
        self._export_roadmap_data(
            roadmap=roadmap,
            output_file=roadmap_file,
        )

        # Export sprint data for each GitHub project that the scrum teams use
        # to manage their sprints, e.g. HHS/17 and HHS/13
        input_files: list[InputFiles] = []
        for sprint_board in self.config.sprint_projects:
            project = sprint_board.project_number
            sprint_file = str(temp_dir / f"sprint-data-{project}.json")
            # Export data
            self._export_sprint_data(
                sprint_board=sprint_board,
                output_file=sprint_file,
            )
            # Add to file list
            input_files.append(
                InputFiles(
                    roadmap=roadmap_file,
                    sprint=sprint_file,
                ),
            )
        # store transient files for re-use during the transform step
        self._transient_files = input_files

    def transform(self) -> None:
        """Run the transformation step of the ETL pipeline."""
        # Load sprint and roadmap data
        issues = []
        for f in self._transient_files:
            issues.extend(run_transformation_pipeline(files=f))
        self.dataset = GitHubIssues(pd.DataFrame(data=issues))

    def load(self) -> None:
        """Run the load step of the ETL pipeline."""
        self.dataset.to_json(self.config.output_file)

    def _export_roadmap_data(
        self,
        roadmap: RoadmapConfig,
        output_file: str,
    ) -> None:
        """Export data from the roadmap project."""
        # Log the export step
        logger.info(
            "Exporting roadmap data from %s/%d",
            roadmap.owner,
            roadmap.project_number,
        )
        # Export the data
        github.export_roadmap_data(
            client=self.client,
            owner=roadmap.owner,
            project=roadmap.project_number,
            quad_field=roadmap.quad_field,
            pillar_field=roadmap.pillar_field,
            output_file=output_file,
        )

    def _export_sprint_data(
        self,
        sprint_board: SprintBoardConfig,
        output_file: str,
    ) -> None:
        """Export data from a sprint board project."""
        logger.info(
            "Exporting sprint data from %s/%s",
            sprint_board.owner,
            sprint_board.project_number,
        )
        github.export_sprint_data(
            client=self.client,
            owner=sprint_board.owner,
            project=sprint_board.project_number,
            sprint_field=sprint_board.sprint_field,
            points_field=sprint_board.points_field,
            output_file=output_file,
        )


# ===============================================================
# Transformation helper functions
# ===============================================================


def run_transformation_pipeline(files: InputFiles) -> list[dict]:
    """Load data from input files and apply transformations."""
    # Log the current sprint for which we're running the transformations
    logger.info("Running transformations for sprint: %s", files.sprint)
    # Load sprint and roadmap data
    sprint_data_in = load_json_file(files.sprint)
    roadmap_data_in = load_json_file(files.roadmap)
    # Populate a lookup table with this data
    lookup: dict = {}
    lookup = populate_issue_lookup_table(lookup, roadmap_data_in)
    lookup = populate_issue_lookup_table(lookup, sprint_data_in)
    # Flatten and write issue level data to output file
    return flatten_issue_data(lookup)


def populate_issue_lookup_table(
    lookup: dict[str, IssueMetadata],
    issues: list[dict],
) -> dict[str, IssueMetadata]:
    """Populate a lookup table that maps issue URLs to their issue type and parent."""
    for i, issue in enumerate(issues):
        try:
            entry = IssueMetadata.model_validate(issue)
        except ValidationError as err:
            logger.error("Error parsing row %d, skipped.", i)  # noqa: TRY400
            logger.debug("Error: %s", err)
            continue
        lookup[entry.issue_url] = entry
    return lookup


def get_parent_with_type(
    child_url: str,
    lookup: dict[str, IssueMetadata],
    type_wanted: IssueType,
) -> IssueMetadata | None:
    """
    Traverse the lookup table to find an issue's parent with a specific type.

    This is useful if we have multiple nested issues, and we want to find the
    top level deliverable or epic that a given task or bug is related to.
    """
    # Get the initial child issue and its parent (if applicable) from the URL
    child = lookup.get(child_url)
    if not child:
        err = f"Lookup doesn't contain issue with url: {child_url}"
        raise ValueError(err)
    if not child.issue_parent:
        return None

    # Travel up the issue hierarchy until we:
    #  - Find a parent issue with the desired type
    #  - Get to an issue without a parent
    #  - Have traversed 5 issues (breaks out of issue cycles)
    max_traversal = 5
    parent_url = child.issue_parent
    for _ in range(max_traversal):
        parent = lookup.get(parent_url)
        # If no parent is found, return None
        if not parent:
            return None
        # If the parent matches the desired type, return it
        if IssueType(parent.issue_type) == type_wanted:
            return parent
        # If the parent doesn't have a its own parent, return None
        if not parent.issue_parent:
            return None
        # Otherwise update the parent_url to "grandparent" and continue
        parent_url = parent.issue_parent

    # Return the URL of the parent deliverable (or None)
    return None


def flatten_issue_data(lookup: dict[str, IssueMetadata]) -> list[dict]:
    """Flatten issue data and inherit data from parent epic an deliverable."""
    result: list[dict] = []
    for issue in lookup.values():
        # If the issue is a deliverable or epic, move to the next one
        if IssueType(issue.issue_type) in [IssueType.DELIVERABLE, IssueType.EPIC]:
            continue

        # Get the parent deliverable, if the issue has one
        deliverable = get_parent_with_type(
            child_url=issue.issue_url,
            lookup=lookup,
            type_wanted=IssueType.DELIVERABLE,
        )
        if deliverable:
            # Set deliverable metadata
            issue.deliverable_title = deliverable.issue_title
            issue.deliverable_url = deliverable.issue_url
            issue.deliverable_pillar = deliverable.deliverable_pillar
            issue.deliverable_status = deliverable.issue_status
            # Set quad metadata
            issue.quad_id = deliverable.quad_id
            issue.quad_name = deliverable.quad_name
            issue.quad_start = deliverable.quad_start
            issue.quad_end = deliverable.quad_end
            issue.quad_length = deliverable.quad_length

        # Get the parent epic, if the issue has one
        epic = get_parent_with_type(
            child_url=issue.issue_url,
            lookup=lookup,
            type_wanted=IssueType.EPIC,
        )
        if epic:
            issue.epic_title = epic.issue_title
            issue.epic_url = epic.issue_url

        # Add the issue to the results
        result.append(issue.model_dump())

    # Return the results
    return result
