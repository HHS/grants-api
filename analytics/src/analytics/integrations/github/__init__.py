"""Export data from GitHub."""

__all__ = [
    "GitHubGraphqlClient",
    "export_roadmap_data",
    "export_sprint_data",
]

from analytics.integrations.github.client import GitHubGraphqlClient
from analytics.integrations.github.main import (
    export_roadmap_data,
    export_sprint_data,
)
