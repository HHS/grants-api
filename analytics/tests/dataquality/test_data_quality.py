"""Test snapshot comparison to known output."""

from pathlib import Path
from unittest import mock

from analytics.etl.github import GitHubProjectConfig, GitHubProjectETL
from analytics.etl.utils import load_config
from analytics.integrations.github.client import GitHubGraphqlClient

from tests.dataquality.inputs.roadmap import mock_graphql_roadmap_data
from tests.dataquality.inputs.sprintboards import mock_graphql_sprintboard_data


def test_roadmap_snapshot(snapshot):
    """Compare pipeline to pre-committed snapshot."""
    test_data = output()

    assert test_data == snapshot()

@mock.patch.object(GitHubGraphqlClient,
                "execute_paginated_query",
             mock_graphql_roadmap_data)
def output() -> list[dict]|None:
    """Call the new pipeline code to be used for comparison."""
    config_path = Path("config/github-projects.json")
    config = load_config(config_path, GitHubProjectConfig)

    # validate effective_date

    return GitHubProjectETL(config).extract_and_transform_in_memory()




def test_sprint_snapshot(snapshot):
    """Compare pipeline for sprintboard snapshot."""
    assert sprint_board_output == snapshot


@mock.patch.object(GitHubGraphqlClient,
                "execute_paginated_query",
             mock_graphql_sprintboard_data)
def sprint_board_output() -> list[dict]|None:
    """Call the new pipeline code to be used for comparison."""
    config_path = Path("config/github-projects.json")
    config = load_config(config_path, GitHubProjectConfig)

    # validate effective_date

    return GitHubProjectETL(config).extract_and_transform_in_memory()
