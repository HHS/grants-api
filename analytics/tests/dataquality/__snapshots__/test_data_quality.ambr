# serializer version: 1
# name: test_roadmap_snapshot
  list([
    dict({
      'deliverable_pillar': 'ðŸ”Ž SimplerFind',
      'deliverable_status': 'Done',
      'deliverable_title': 'Search *',
      'deliverable_url': 'https://github.com/HHS/simpler-grants-gov/issues/2200',
      'epic_title': 'Search API Engagement',
      'epic_url': 'https://github.com/HHS/simpler-grants-gov/issues/2719',
      'issue_closed_at': '2024-11-12',
      'issue_is_closed': True,
      'issue_opened_at': '2024-11-07',
      'issue_parent': 'https://github.com/HHS/simpler-grants-gov/issues/2719',
      'issue_points': None,
      'issue_state': 'closed',
      'issue_status': 'Done',
      'issue_title': 'Draft email content for participants',
      'issue_type': 'Task',
      'issue_url': 'https://github.com/HHS/simpler-grants-gov/issues/2763',
      'project_number': 13,
      'project_owner': 'HHS',
      'quad_end': '2025-01-09',
      'quad_id': '93412e1c',
      'quad_length': 122.0,
      'quad_name': 'Quad 1.1',
      'quad_start': '2024-09-09',
      'sprint_end': None,
      'sprint_id': None,
      'sprint_length': None,
      'sprint_name': None,
      'sprint_start': None,
    }),
    dict({
      'deliverable_pillar': None,
      'deliverable_status': None,
      'deliverable_title': None,
      'deliverable_url': None,
      'epic_title': None,
      'epic_url': None,
      'issue_closed_at': '2024-11-25',
      'issue_is_closed': True,
      'issue_opened_at': '2024-11-13',
      'issue_parent': None,
      'issue_points': None,
      'issue_state': 'closed',
      'issue_status': 'Done',
      'issue_title': '[DRAFT] Save and Subscribe to a Search - Not Ready for Feedback',
      'issue_type': 'Task',
      'issue_url': 'https://github.com/HHS/simpler-grants-gov/issues/2820',
      'project_number': 13,
      'project_owner': 'HHS',
      'quad_end': None,
      'quad_id': None,
      'quad_length': None,
      'quad_name': None,
      'quad_start': None,
      'sprint_end': None,
      'sprint_id': None,
      'sprint_length': None,
      'sprint_name': None,
      'sprint_start': None,
    }),
    dict({
      'deliverable_pillar': 'ðŸ”Ž SimplerFind',
      'deliverable_status': 'Done',
      'deliverable_title': 'Search *',
      'deliverable_url': 'https://github.com/HHS/simpler-grants-gov/issues/2200',
      'epic_title': 'Search API Engagement',
      'epic_url': 'https://github.com/HHS/simpler-grants-gov/issues/2719',
      'issue_closed_at': '2024-11-12',
      'issue_is_closed': True,
      'issue_opened_at': '2024-11-07',
      'issue_parent': 'https://github.com/HHS/simpler-grants-gov/issues/2719',
      'issue_points': None,
      'issue_state': 'closed',
      'issue_status': 'Done',
      'issue_title': 'Draft email content for participants',
      'issue_type': 'Task',
      'issue_url': 'https://github.com/HHS/simpler-grants-gov/issues/2763',
      'project_number': 17,
      'project_owner': 'HHS',
      'quad_end': '2025-01-09',
      'quad_id': '93412e1c',
      'quad_length': 122.0,
      'quad_name': 'Quad 1.1',
      'quad_start': '2024-09-09',
      'sprint_end': None,
      'sprint_id': None,
      'sprint_length': None,
      'sprint_name': None,
      'sprint_start': None,
    }),
    dict({
      'deliverable_pillar': None,
      'deliverable_status': None,
      'deliverable_title': None,
      'deliverable_url': None,
      'epic_title': None,
      'epic_url': None,
      'issue_closed_at': '2024-11-25',
      'issue_is_closed': True,
      'issue_opened_at': '2024-11-13',
      'issue_parent': None,
      'issue_points': None,
      'issue_state': 'closed',
      'issue_status': 'Done',
      'issue_title': '[DRAFT] Save and Subscribe to a Search - Not Ready for Feedback',
      'issue_type': 'Task',
      'issue_url': 'https://github.com/HHS/simpler-grants-gov/issues/2820',
      'project_number': 17,
      'project_owner': 'HHS',
      'quad_end': None,
      'quad_id': None,
      'quad_length': None,
      'quad_name': None,
      'quad_start': None,
      'sprint_end': None,
      'sprint_id': None,
      'sprint_length': None,
      'sprint_name': None,
      'sprint_start': None,
    }),
  ])
# ---
# name: test_sprint_snapshot
  'sprint_board_output() -> list[dict] | None'
# ---
